### ISSUE TITLE
Bug: Select DISTINCT (col) does not return correct result ...

### ISSUE THREAD

**Issue by HeikoOnnebrink** [cockroachdb/cockroach#18586]
Is this a question, feature request, or bug report?

bug

Have you checked our documentation at https://cockroachlabs.com/docs/stable/? If you could not find an answer there, please consider asking your question in our community forum at https://forum.cockroachlabs.com/, as it would benefit other members of our community.

Prefer live chat? Message our engineers on our Gitter channel at https://gitter.im/cockroachdb/cockroach.

**FEATURE REQUEST**

1. Does an issue already exist addressing this request? If yes, please add a :+1: reaction to the existing issue. If not, move on to step 2.

2. Please describe the feature you are requesting, as well as your proposed use case for this feature.

3. Indicate the importance of this issue to you (blocker, must-have, should-have, nice-to-have). Are you currently using any workarounds to address this issue?

**BUG REPORT**

1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems
   running with defaults, this boils down to the output of

     grep -F '[config]' cockroach-data/logs/cockroach.log

   When log files are not available, supply the output of `cockroach version`
   and all flags/environment variables passed to `cockroach start` instead.

2. Please describe the issue you observed:

- What did you do?

- What did you expect to see?

- What did you see instead?

W170915 11:12:23.488359 1 cli/start.go:650  RUNNING IN INSECURE MODE!

- Your cluster is open for any client that can access <all your IP addresses>.
- Any user, even root, can log in without providing a password.
- Any user, connecting as root, can read or write any data in your cluster.
- There is no network encryption nor authentication, and thus no confidentiality.

Check out how to secure your cluster: https://www.cockroachlabs.com/docs/stable/secure-a-cluster.html

```
I170915 11:12:23.489434 1 server/config.go:303  available memory from cgroups (8.0 EiB) exceeds system memory 3.9 GiB, using system memory
W170915 11:12:23.489487 1 cli/start.go:570  Using the default setting for --cache (128 MiB).
  A significantly larger value is usually needed for good performance.
  If you have a dedicated server a reasonable setting is --cache=25% (989 MiB).
I170915 11:12:23.489513 1 cli/start.go:658  CockroachDB CCL be60cc5-dirty (linux amd64, built 2017/09/15 11:04:19, go1.8.3)
I170915 11:12:23.590373 1 server/config.go:303  available memory from cgroups (8.0 EiB) exceeds system memory 3.9 GiB, using system memory
I170915 11:12:23.590420 1 server/config.go:417  system total memory: 3.9 GiB
I170915 11:12:23.590633 1 server/config.go:419  server configuration:
max offset                500000000
cache size                128 MiB
SQL memory pool size      128 MiB
scan interval             10m0s
scan max idle time        200ms
metrics sample interval   10s
event log enabled         true
linearizable              false
```

have some fresh created 9 node cluster across 3 dc with custom build from 1.1 branch.
Looks like SELECT DISTINCT .. has some issues : 

have some table requests with below structure;

```
| Field | Type | Null | Default | Indices |
| requestkey | INT | false | unique_rowid() | {"primary","requests_requeststate_dcname_id_idx"} |
| requesttype | STRING | false | NULL | {} |
| requesttime | TIMESTAMP WITH TIME ZONE | true | NULL | {} |
| requeststate | STRING | true | NULL | {"requests_requeststate_dcname_id_idx"} |
| trace | STRING | true | NULL | {} |
| laststep | STRING | true | NULL | {} |
| lasterror | STRING | true | NULL | {} |
| resourcetype | STRING | false | NULL | {} |
| dcname | STRING | true | NULL | {"requests_requeststate_dcname_id_idx"} |
| id | BYTES | false | NULL | {"requests_requeststate_dcname_id_idx"} |
| body | STRING | true | NULL | {} |
```

Data distribution looks like this

```
select resourceType,count(1) from requests group by resourceType;
+--------------+----------+
| resourcetype | count(1) |
+--------------+----------+
| INTERNAL | 315 |
| DB | 64 |
| JOB | 35 |
```

execute

```
select distinct resourceType from requests;
+--------------+
| resourcetype |
+--------------+
| INTERNAL |
| DB |
| DB |
| DB |
| JOB |
| JOB |
| INTERNAL |
+--------------+
(7 rows)
```

Got 7 rows, expected 3 

Executing 

```
select distinct 'X'||resourceType||'X' from requests;
+------------------------------+
| ('X' || resourcetype) || 'X' |
+------------------------------+
| XINTERNALX |
| XDBX |
| XJOBX |
+------------------------------+
(3 rows)
```

Was not yet able to create a reproducible testcase.
A short test works generally fine for same type of query.

```
apiuserftx@10.96.214.109:26257/restapiftx> create table test (a string);
CREATE TABLE
apiuserftx@10.96.214.109:26257/restapiftx> INSERT INTO test VALUES ('A');
INSERT 1
apiuserftx@10.96.214.109:26257/restapiftx> INSERT INTO test VALUES ('A');
INSERT 1
apiuserftx@10.96.214.109:26257/restapiftx> INSERT INTO test VALUES ('B');
INSERT 1
apiuserftx@10.96.214.109:26257/restapiftx> INSERT INTO test VALUES ('B');
INSERT 1
apiuserftx@10.96.214.109:26257/restapiftx> SELECT * FROM test;
+---+
| a |
+---+
| A |
| A |
| B |
| B |
+---+
(4 rows)
apiuserftx@10.96.214.109:26257/restapiftx> SELECT DISTINCT a FROM test;
+---+
| a |
+---+
| A |
| B |
+---+
(2 rows)
```

**Comment by HeikoOnnebrink** [comment]
in case access to cluster is wanted to have a look I could provide a WebEx session ..

**Comment by arjunravinarayan** [comment]
Hi @HeikoOnnebrink, thank you very much for the report. I'm sorry that there is a semantic bug in the query execution. I will try and fix this as rapidly as possible. Right now, I'm looking for a few bits of information to debug this further. Can you please send me the result of running the following queries on your cluster?

1. `SHOW CREATE TABLE requests`
2. `SHOW INDEXES FROM requests`
3. `EXPLAIN (SELECT DISTINCT resourceType FROM requests)`
4. `EXPLAIN (DISTSQL) (SELECT DISTINCT resourceType FROM requests)`

I would also like you to do two things:
1. See if the problem recurs when you turn off Distributed SQL execution (run `SET CLUSTER SETTING sql.defaults.distsql = off;` and then run the query).
2. You can then reset Distributed SQL execution to the default with `SET CLUSTER SETTING sql.defaults.distsql = auto;`.

For an explanation of why I want this information, I'm trying to understand:
1) What orderings are available on the `requests` column when running this query.
2) What orderings the query is using.

Thank you.

**Comment by HeikoOnnebrink** [comment]
here the output as requested ...
cheers
Heiko

1.)
+----------+----------------------------------------------------------------------------------------------------------------------------------------+
|  Table   |                                                              CreateTable                                                               |
+----------+----------------------------------------------------------------------------------------------------------------------------------------+
| requests | CREATE TABLE requests (␤                                                                                                               |
|          |     requestkey INT NOT NULL DEFAULT unique_rowid(),␤                                                                                   |
|          |     requesttype STRING NOT NULL,␤                                                                                                      |
|          |     requesttime TIMESTAMP WITH TIME ZONE NULL,␤                                                                                        |
|          |     requeststate STRING NULL,␤                                                                                                         |
|          |     trace STRING NULL,␤                                                                                                                |
|          |     laststep STRING NULL,␤                                                                                                             |
|          |     lasterror STRING NULL,␤                                                                                                            |
|          |     resourcetype STRING NOT NULL,␤                                                                                                     |
|          |     dcname STRING NULL,␤                                                                                                               |
|          |     id BYTES NOT NULL,␤                                                                                                                |
|          |     body STRING NULL,␤                                                                                                                 |
|          |     CONSTRAINT "primary" PRIMARY KEY (requestkey ASC),␤                                                                                |
|          |     INDEX requests_requeststate_dcname_id_idx (requeststate ASC, dcname ASC, id ASC) STORING (requestkey),␤                            |
|          |     FAMILY "primary" (requestkey, requesttype, requesttime, requeststate, trace, laststep, lasterror, resourcetype, dcname, id, body)␤ |
|          | )                                                                                                                                      |
+----------+----------------------------------------------------------------------------------------------------------------------------------------+

2.)
+----------+-------------------------------------+--------+-----+--------------+-----------+---------+----------+
|  Table   |                Name                 | Unique | Seq |    Column    | Direction | Storing | Implicit |
+----------+-------------------------------------+--------+-----+--------------+-----------+---------+----------+
| requests | primary                             | true   |   1 | requestkey   | ASC       | false   | false    |
| requests | requests_requeststate_dcname_id_idx | false  |   1 | requeststate | ASC       | false   | false    |
| requests | requests_requeststate_dcname_id_idx | false  |   2 | dcname       | ASC       | false   | false    |
| requests | requests_requeststate_dcname_id_idx | false  |   3 | id           | ASC       | false   | false    |
| requests | requests_requeststate_dcname_id_idx | false  |   4 | requestkey   | N/A       | true    | false    |
+----------+-------------------------------------+--------+-----+--------------+-----------+---------+----------+

3.)
+-------+----------+-------+------------------+
| Level |   Type   | Field |   Description    |
+-------+----------+-------+------------------+
|     0 | distinct |       |                  |
|     1 | render   |       |                  |
|     2 | scan     |       |                  |
|     2 |          | table | requests@primary |
|     2 |          | spans | ALL              |
+-------+----------+-------+------------------+

4.)
+-----------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Automatic |                                                                                                                                                                                        URL                                                                                                                                                                                         |                                                                                                                                                                                                                                                                                                                       JSON                                                                                                                                                                                                                                                                                                                       |
+-----------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| true      | https://cockroachdb.github.io/distsqlplan/decode.html?eJyUkT1rwzAQhvf-ivLOKthOh6LJQ5csSQndigfVOoLA8bm6E7QE__dia2hciNuM9_E8rzid0bOnnTuRwL6hhMEjGoMhcksiHKd2Xtr6T9jCIPRD0qndGLQcCfYMDdoRLHb8wAMMPKkL3bw0GnDSH0TUHQl2M5oLbbmufXXvHR3IeYoLOYYYTi5-1ZE-EokKDPZJ7X39hGvB5S3Bz0E09K0uU-vyqr1a2P-41oFk4F7oXwcrxsaA_JHyjwin2NJL5HaOyeV-5uaGJ9E83eRi2-fR9MBLuFyFq3W4WoWLX3Az3n0HAAD__4q4zsA= | {"nodeNames":["1","4"],"processors":[{"nodeIdx":0,"inputs":[],"core":{"title":"No-op","details":[]},"outputs":[],"stage":3},{"nodeIdx":1,"inputs":[],"core":{"title":"TableReader","details":["primary@requests","Out: @8"]},"outputs":[],"stage":1},{"nodeIdx":1,"inputs":[],"core":{"title":"Distinct","details":["@1"]},"outputs":[],"stage":2},{"nodeIdx":0,"inputs":[],"core":{"title":"Response","details":[]},"outputs":[],"stage":0}],"edges":[{"sourceProc":0,"sourceOutput":0,"destProc":3,"destInput":0},{"sourceProc":1,"sourceOutput":0,"destProc":2,"destInput":0},{"sourceProc":2,"sourceOutput":0,"destProc":0,"destInput":0}]}␤ |
|           |                                                                                                                                                                                                                                                                                                                                                                                    |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
+-----------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

.. and the distsql test ...

root@10.96.214.109:26257/restapi> SET CLUSTER SETTING sql.defaults.distsql = off;
SET
root@10.96.214.109:26257/restapi> select distinct resourceType from requests;
+--------------+
| resourcetype |
+--------------+
| INTERNAL     |
| JOB          |
| DB           |
| DB           |
| INTERNAL     |
| DB           |
| JOB          |
| INTERNAL     |
+--------------+
(8 rows)
root@10.96.214.109:26257/restapi> SET CLUSTER SETTING sql.defaults.distsql = 'auto';
SET
root@10.96.214.109:26257/restapi> SELECT DISTINCT resourcetype FROM requests;
+--------------+
| resourcetype |
+--------------+
| INTERNAL     |
| JOB          |
| DB           |
| DB           |
| INTERNAL     |
| DB           |
| JOB          |
| INTERNAL     |
+--------------+
(8 rows)

**Comment by arjunravinarayan** [comment]
I've edited your issue for formatting. I was mistaken in giving you directions for disabling Distributed SQL execution. The command I gave you only sets the DistSQL mode for *new* sessions. Can you please run:

`SET DISTSQL = OFF`, and then rerun the query?

**Comment by HeikoOnnebrink** [comment]
did the test and seems to confirm your assumption 🥇 

```
select distinct resourceType from requests;
+--------------+
| resourcetype |
+--------------+
| INTERNAL     |
| JOB          |
| DB           |
| DB           |
| INTERNAL     |
| DB           |
| JOB          |
| INTERNAL     |
+--------------+
(8 rows)
apiuser@10.96.214.109:26257/restapi> SET DISTSQL = OFF;
SET
apiuser@10.96.214.109:26257/restapi> select distinct resourceType from requests;
+--------------+
| resourcetype |
+--------------+
| INTERNAL     |
| JOB          |
| DB           |
+--------------+
(3 rows)
```

**Comment by HeikoOnnebrink** [comment]
`explain (verbose) (select distinct resourcetype from requests); 
+-------+----------+----------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+
| Level |   Type   |  Field   |          Description          |                                                                                                  Columns                                                                                                   | Ordering |
+-------+----------+----------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+
|     0 | distinct |          |                               | (resourcetype)                                                                                                                                                                                             |          |
|     1 | render   |          |                               | (resourcetype)                                                                                                                                                                                             |          |
|     1 |          | render 0 | restapi.requests.resourcetype |                                                                                                                                                                                                            |          |
|     2 | scan     |          |                               | (requestkey[omitted], requesttype[omitted], requesttime[omitted], requeststate[omitted], trace[omitted], laststep[omitted], lasterror[omitted], resourcetype, dcname[omitted], id[omitted], body[omitted]) |          |
|     2 |          | table    | requests@primary              |                                                                                                                                                                                                            |          |
|     2 |          | spans    | ALL                           |                                                                                                                                                                                                            |          |
+-------+----------+----------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+
`

**Comment by arjunravinarayan** [comment]
@HeikoOnnebrink thank you very much for the bug report. I have identified the source of the bug, and have prepared a patch to fix it(#18702). Once that patch is merged, I'll make sure it is added to the next release alpha.

The reason I was unable to reproduce the bug was that it's due to the varying number of NULLs that precede the `resourceType` column: until I had your data dump, I did not even try that possibility. Thank you once again.

**Comment by HeikoOnnebrink** [comment]
@arjunravinarayan thanks for the update and really detailed explanation. As I am running custom builds  all I need to know is to which branch you merged the fix, 1.1 , Master, other ... ? I will than build a new bin and digest the change directly.. cheers, Heiko

**Comment by arjunravinarayan** [comment]
@HeikoOnnebrink, it's now merged into both `master` and the `release-1.1` branch. I recommend you download from the `release-1.1` branch, as `master` is more experimental.

You should be able to either build it locally, or download Linux binaries using `curl -L https://edge-binaries.cockroachdb.com/cockroach/cockroach.linux-gnu-amd64.release-1.1`. By late next week the latest [testing release](curl -L https://edge-binaries.cockroachdb.com/cockroach/cockroach.linux-gnu-amd64.release-1.1) will contain this patch as well.

**Comment by vivekmenezes** [comment]
vibrassimo!!!!! Great find
